File: lib/python-sdk/examples/transforms.py
#!/usr/bin/env python3
"""Bidirectional transform PoC — plugin transformation interface.

Demonstrates source (grants.gov) → CommonGrants and CommonGrants → source
bidirectional transformations using the grants.gov sample plugin.

Requires generated schemas (examples/plugins/grants_gov/generated/).
Generate them first (from lib/python-sdk/):
    poetry run python -m common_grants_sdk.extensions.generate --plugin examples/plugins/grants_gov
Or run all plugins at once:
    make plugins

Then run (from lib/python-sdk/):
    poetry run python examples/transforms.py
"""

from __future__ import annotations

import json
from typing import Any

# When run as `poetry run python examples/transforms.py`, Python automatically
# adds the script's directory (examples/) to sys.path. Import from there using
# the `plugins.` prefix (not `examples.plugins.`) — the `examples.` prefix only
# works in -c or interactive contexts where lib/python-sdk/ is sys.path[0].
from plugins.grants_gov import grants_gov as plugin
from plugins.grants_gov.generated.schemas import Opportunity

from common_grants_sdk.extensions import build_transforms
from common_grants_sdk.utils.transformation import get_from_path

# ---------------------------------------------------------------------------
# Sample grants.gov source data
# ---------------------------------------------------------------------------

SOURCE_DATA: dict[str, Any] = {
    "data": {
        "agency_name": "Department of Examples",
        "created_at": "2025-01-15T09:00:00Z",
        "last_modified_at": "2025-04-01T12:30:00Z",
        "opportunity_description": "Funding to advance research into conservation techniques for endangered ecosystems.",
        "opportunity_id": 12345,
        "opportunity_number": "ABC-123-XYZ-001",
        "opportunity_status": "posted",
        "opportunity_title": "Research into conservation techniques",
        "opportunity_uuid": "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
        "priority_score_str": "75",
        "summary": {
            "applicant_types": ["state_governments"],
            "archive_date": "2025-05-01",
            "award_ceiling": 100000,
            "award_floor": 10000,
            "forecasted_award_date": "2025-09-01",
            "forecasted_close_date": "2025-07-15",
            "forecasted_post_date": "2025-05-01",
        },
    }
}


# ---------------------------------------------------------------------------
# Custom handlers: join_fields and split_field
#
# join_fields concatenates multiple source field values with a configurable
# separator. Mapping spec: {"join": {"fields": ["a.b", "c.d"], "sep": " — "}}
#
# split_field is the inverse: it splits a single field on a separator and
# returns the element at the given index.
# Mapping spec: {"split": {"field": "label", "sep": " — ", "index": 0}}
# ---------------------------------------------------------------------------


def join_fields(data: dict[str, Any], spec: dict[str, Any]) -> str | None:
    """Custom handler that joins multiple field values with a separator."""
    sep = spec.get("sep", " ")
    parts = [get_from_path(data, path) for path in spec.get("fields", [])]
    values = [str(p) for p in parts if p is not None]
    return sep.join(values) if values else None


def split_field(data: dict[str, Any], spec: dict[str, Any]) -> str | None:
    """Custom handler that splits a field value and returns the element at index."""
    value = get_from_path(data, spec.get("field", ""))
    if value is None:
        return None
    parts = str(value).split(spec.get("sep", " "))
    index = spec.get("index", 0)
    return parts[index] if index < len(parts) else None


# Transform that uses the custom handlers and validates output against the generated
# Opportunity model. common_model=Opportunity (from generated/schemas.py) ensures
# model_validate runs against the extended class with typed custom fields
# (legacyId, agencyName, applicantTypes), not just the base OpportunityBase.
to_common_with_custom, from_common_with_custom = build_transforms(
    to_common_mapping={
        "id": {"field": "data.opportunity_uuid"},
        "title": {"field": "data.opportunity_title"},
        "description": {"field": "data.opportunity_description"},
        "createdAt": {"field": "data.created_at"},
        "lastModifiedAt": {"field": "data.last_modified_at"},
        "status": {
            "value": {
                "match": {
                    "field": "data.opportunity_status",
                    "case": {
                        "posted": "open",
                        "archived": "closed",
                        "forecasted": "forecasted",
                    },
                    "default": "custom",
                }
            },
        },
        "label": {
            "join": {
                "fields": ["data.opportunity_number", "data.opportunity_title"],
                "sep": " — ",
            }
        },
        "customFields": {
            "legacyId": {
                "value": {"field": "data.opportunity_id"},
            },
            "agencyName": {
                "value": {"field": "data.agency_name"},
            },
            "applicantTypes": {
                "value": {"field": "data.summary.applicant_types"},
            },
        },
    },
    from_common_mapping={
        "data": {
            # label is produced by the join handler above but gets dropped by
            # model_validate (it is not a CG field), so from_common maps directly
            # from the standard CG title field instead.
            "opportunity_title": {"field": "title"},
        }
    },
    handlers={"join": join_fields, "split": split_field},
    common_model=Opportunity,
)


def _section(title: str) -> None:
    print(f"\n{'=' * 60}")
    print(title)
    print("=" * 60)


def main() -> None:
    opp = plugin.schemas.Opportunity

    _section("SOURCE DATA (grants.gov format)")
    print(json.dumps(SOURCE_DATA, indent=2))

    # --- to_common: grants.gov → CommonGrants ---
    assert opp.to_common is not None
    cg_result = opp.to_common(SOURCE_DATA)

    _section("to_common: grants.gov → CommonGrants")
    if cg_result.errors:
        print(f"ERRORS ({len(cg_result.errors)}):")
        for err in cg_result.errors:
            print(f"  [path={err.path}] {err}")
    else:
        print("Errors: none")
    print("\nResult:")
    print(json.dumps(cg_result.result, indent=2))

    # --- from_common: CommonGrants → grants.gov ---
    assert opp.from_common is not None
    native_result = opp.from_common(cg_result.result)

    _section("from_common: CommonGrants → grants.gov")
    if native_result.errors:
        print(f"ERRORS ({len(native_result.errors)}):")
        for err in native_result.errors:
            print(f"  [path={err.path}] {err}")
    else:
        print("Errors: none")
    print("\nResult:")
    print(json.dumps(native_result.result, indent=2))

    # --- Roundtrip comparison ---
    # Note: SOURCE_DATA contains fields not covered by the mappings (agency_name,
    # opportunity_id, etc.). Those fields are intentionally absent from the roundtrip
    # output — the mapping layer is selective by design.
    _section("ROUNDTRIP CHECK")
    checks = [
        (
            "title",
            SOURCE_DATA["data"]["opportunity_title"],
            native_result.result.get("data", {}).get("opportunity_title"),
        ),
        (
            "status",
            SOURCE_DATA["data"]["opportunity_status"],
            native_result.result.get("data", {}).get("opportunity_status"),
        ),
        (
            "award_floor",
            SOURCE_DATA["data"]["summary"]["award_floor"],
            native_result.result.get("data", {}).get("summary", {}).get("award_floor"),
        ),
        (
            "award_ceiling",
            SOURCE_DATA["data"]["summary"]["award_ceiling"],
            native_result.result.get("data", {})
            .get("summary", {})
            .get("award_ceiling"),
        ),
        (
            "priority_score_str",
            SOURCE_DATA["data"]["priority_score_str"],
            native_result.result.get("data", {}).get("priority_score_str"),
        ),
    ]
    all_pass = True
    for field, original, roundtripped in checks:
        ok = original == roundtripped
        if not ok:
            all_pass = False
        status = "PASS" if ok else "FAIL"
        print(f"  [{status}] {field}: {original!r} -> {roundtripped!r}")

    print(
        f"\nRoundtrip result ({len(checks)} mapped fields checked; unmapped fields dropped by design): {'ALL PASS' if all_pass else 'SOME FIELDS DIFFER'}"
    )

    # --- Custom handler + model_validate demo ---
    _section(
        "CUSTOM HANDLER + MODEL VALIDATE DEMO (join / split / extended Opportunity)"
    )
    print("Custom handlers: join, split")
    print("common_model: generated Opportunity (with typed customFields)\n")

    custom_cg = to_common_with_custom(SOURCE_DATA)

    if custom_cg.errors:
        print(f"ERRORS ({len(custom_cg.errors)}):")
        for err in custom_cg.errors:
            print(f"  [path={err.path}] {err}")
    else:
        print("Validation: PASS — result is a typed Opportunity instance")
        assert isinstance(custom_cg.result, Opportunity)
        opp_instance = custom_cg.result
        print(f"\n  title:       {opp_instance.title}")
        print(f"  id:          {opp_instance.id}")
        print(f"  status:      {opp_instance.status.value}")
        if opp_instance.custom_fields:
            cf = opp_instance.custom_fields
            print("\n  customFields (typed):")
            if cf.legacy_id:
                print(
                    f"    legacyId.value:        {cf.legacy_id.value!r}  ({type(cf.legacy_id.value).__name__})"
                )
            if cf.agency_name:
                print(
                    f"    agencyName.value:      {cf.agency_name.value!r}  ({type(cf.agency_name.value).__name__})"
                )
            if cf.applicant_types:
                print(
                    f"    applicantTypes.value:  {cf.applicant_types.value!r}  ({type(cf.applicant_types.value).__name__})"
                )

    custom_native = from_common_with_custom(
        custom_cg.result if not custom_cg.errors else {}
    )
    orig_title = SOURCE_DATA["data"]["opportunity_title"]
    rt_title = custom_native.result.get("data", {}).get("opportunity_title")
    print(
        f"\n  [{'PASS' if orig_title == rt_title else 'FAIL'}] opportunity_title: {orig_title!r} -> {rt_title!r}"
    )

    # --- Plugin metadata ---
    _section("PLUGIN METADATA")
    assert plugin.meta is not None
    print(f"name:         {plugin.meta.name}")
    print(f"version:      {plugin.meta.version}")
    print(f"sourceSystem: {plugin.meta.source_system}")
    print(f"capabilities: {plugin.meta.capabilities}")


if __name__ == "__main__":
    main()

File: lib/ts-sdk/examples/transforms.ts
/**
 * Example script demonstrating bidirectional transforms through `definePlugin()`.
 *
 * Runs real data through each authoring path and verifies the round trip. It does
 * not call `buildTransforms()` directly — `definePlugin()` compiles the declarative
 * `mappings` internally. (The schema-only path, plus the schema-only / XOR
 * compile-time guarantees, are covered in `__tests__/extensions/define-plugin.spec.ts`.)
 *
 * Paths shown:
 *   1. Custom fields + declarative `mappings`   → definePlugin compiles them
 *   2. Custom fields + hand-written functions    → `ToCommon` / `FromCommon`
 *   3. Transforms with no custom fields          → base CommonGrants schema
 *   X. `mappings` + functions on one entry       → rejected (mappings XOR functions)
 *
 * Run with: `pnpm example:transforms`
 *
 * @remarks
 * The mappings path validates `toCommon` output against the extended common
 * schema, so date strings are parsed into `Date` objects on the common side.
 * The round-trip checks below therefore assert on fields that survive verbatim
 * (ids, the three-state `source` null), not on the parsed date fields.
 */

import { z } from "zod";

import { CustomFieldType } from "../src/constants";
import {
  definePlugin,
  type FromCommon,
  type ToCommon,
  type TransformResult,
} from "../src/extensions";

// ############################################################################
// Shared setup — source schema, sample data, custom fields
// ############################################################################

// Stands in for a real grants.gov payload schema. In a real plugin this would be
// imported from wherever the source-system types live.
const GrantsGovOpportunity = z.object({
  data: z.object({
    opportunity_uuid: z.string().uuid(),
    opportunity_id: z.number().int(),
    opportunity_title: z.string(),
    opportunity_description: z.string(),
    opportunity_status: z.string(),
    // `source_url: null` below is the publisher actively asserting "doesn't
    // apply" (three-state null). The transforms preserve it as `null` end to end
    // rather than collapsing it to absent.
    source_url: z.string().url().nullish(),
    created_at: z.string(),
    last_modified_at: z.string(),
  }),
});
type GrantsGovSource = z.infer<typeof GrantsGovOpportunity>;

const SOURCE_DATA: GrantsGovSource = {
  data: {
    opportunity_uuid: "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
    opportunity_id: 12345,
    opportunity_title: "Research into conservation techniques",
    opportunity_description: "Funding to advance conservation research.",
    opportunity_status: "posted",
    source_url: null,
    created_at: "2025-01-15T09:00:00Z",
    last_modified_at: "2025-04-01T12:30:00Z",
  },
};

// Declared inline at each call below, but kept here too for the hand-written
// path, which needs `typeof customFields` for its helper-type annotations.
const customFields = {
  legacyId: {
    fieldType: CustomFieldType.integer,
    value: z.number().int(),
    description: "Numeric ID from the legacy system (round-trip preserved).",
  },
} as const;

// ############################################################################
// Path 1 — custom fields + declarative mappings
// ############################################################################

// definePlugin() compiles these mappings into toCommon / fromCommon using the
// built-in handlers (`field`, `match`, `const`). No buildTransforms() call here.
const mappingsPlugin = definePlugin({
  meta: { name: "grants.gov (mappings)", sourceSystem: "grants.gov" },
  schemas: {
    Opportunity: {
      customFields,
      sourceSchema: GrantsGovOpportunity,
      mappings: {
        toCommon: {
          id: { field: "data.opportunity_uuid" },
          title: { field: "data.opportunity_title" },
          description: { field: "data.opportunity_description" },
          // Three-state null: a terminal `null` is preserved as a real null.
          source: { field: "data.source_url" },
          createdAt: { field: "data.created_at" },
          lastModifiedAt: { field: "data.last_modified_at" },
          status: {
            value: {
              match: {
                field: "data.opportunity_status",
                case: { posted: "open", archived: "closed" },
                default: "custom",
              },
            },
          },
          customFields: {
            legacyId: {
              value: { field: "data.opportunity_id" },
              name: "legacyId",
              fieldType: "integer",
            },
          },
        },
        fromCommon: {
          data: {
            opportunity_uuid: { field: "id" },
            opportunity_title: { field: "title" },
            opportunity_description: { field: "description" },
            source_url: { field: "source" },
            created_at: { field: "createdAt" },
            last_modified_at: { field: "lastModifiedAt" },
            opportunity_status: { const: "posted" },
            opportunity_id: { field: "customFields.legacyId.value" },
          },
        },
      },
    },
  },
});

// ############################################################################
// Path 2 — custom fields + hand-written functions
// ############################################################################

// The author annotates the functions with `ToCommon` / `FromCommon`, passing the
// same inputs `definePlugin()` uses — `model`, `sourceSchema`, and the
// `customFields` specs they already have. The SDK resolves the common type from
// those (no need to build or borrow a `commonSchema`). One common type is used
// for both directions: the common date schemas accept either a string or a
// `Date`, so the author builds and returns real `Date` values here, and reads
// `Date` values back in `fromCommon`.
type OpportunityTransform = {
  model: "Opportunity";
  sourceSchema: typeof GrantsGovOpportunity;
  customFields: typeof customFields;
};

// The functions path is the override path for logic the declarative mappings
// can't express — here, a status flag derived in code rather than via `match`.
const toCommon: ToCommon<OpportunityTransform> = source => ({
  result: {
    id: source.data.opportunity_uuid,
    title: source.data.opportunity_title,
    description: source.data.opportunity_description,
    source: source.data.source_url,
    status: { value: source.data.opportunity_status === "posted" ? "open" : "custom" },
    createdAt: new Date(source.data.created_at),
    lastModifiedAt: new Date(source.data.last_modified_at),
    customFields: {
      legacyId: { name: "legacyId", fieldType: "integer", value: source.data.opportunity_id },
    },
  },
  errors: [],
});

const fromCommon: FromCommon<OpportunityTransform> = common => ({
  result: {
    data: {
      opportunity_uuid: common.id,
      opportunity_id: common.customFields?.legacyId?.value ?? 0,
      opportunity_title: common.title,
      opportunity_description: common.description,
      opportunity_status: "posted",
      source_url: common.source ?? null,
      // `createdAt` / `lastModifiedAt` arrive as `Date`; render back to strings.
      created_at: common.createdAt.toISOString(),
      last_modified_at: common.lastModifiedAt.toISOString(),
    },
  },
  errors: [],
});

const functionsPlugin = definePlugin({
  meta: { name: "grants.gov (functions)", sourceSystem: "grants.gov" },
  schemas: {
    Opportunity: { customFields, sourceSchema: GrantsGovOpportunity, toCommon, fromCommon },
  },
});

// ############################################################################
// Path 3 — transforms with no custom fields
// ############################################################################

// Omit `customFields`; the common schema is the base CommonGrants Opportunity.
const noCustomFieldsPlugin = definePlugin({
  meta: { name: "grants.gov (no custom fields)", sourceSystem: "grants.gov" },
  schemas: {
    Opportunity: {
      sourceSchema: GrantsGovOpportunity,
      mappings: {
        toCommon: {
          id: { field: "data.opportunity_uuid" },
          title: { field: "data.opportunity_title" },
          description: { field: "data.opportunity_description" },
          source: { field: "data.source_url" },
          createdAt: { field: "data.created_at" },
          lastModifiedAt: { field: "data.last_modified_at" },
          status: {
            value: {
              match: {
                field: "data.opportunity_status",
                case: { posted: "open", archived: "closed" },
                default: "custom",
              },
            },
          },
        },
        fromCommon: {
          data: {
            opportunity_uuid: { field: "id" },
            opportunity_title: { field: "title" },
            opportunity_description: { field: "description" },
            source_url: { field: "source" },
            created_at: { field: "createdAt" },
            last_modified_at: { field: "lastModifiedAt" },
            opportunity_status: { const: "posted" },
          },
        },
      },
    },
  },
});

// ############################################################################
// Path X — mappings + functions on one entry is rejected (XOR)
// ############################################################################

// Providing both `mappings` and explicit callables is a compile error and a
// runtime error. Wrapped here so the runtime backstop can be demonstrated
// without aborting the script.
function demonstrateXorIsRejected(): void {
  try {
    definePlugin({
      schemas: {
        Opportunity: {
          sourceSchema: GrantsGovOpportunity,
          mappings: { toCommon: {}, fromCommon: {} },
          // @ts-expect-error — mappings XOR functions: cannot provide both on one entry
          toCommon,
        },
      },
    });
    fail("path X: expected definePlugin to reject mappings + functions");
  } catch {
    console.log("path X (XOR): mappings + functions rejected at runtime, too");
  }
}

// ############################################################################
// Run the round trips and report
// ############################################################################

function fail(message: string): never {
  console.error(message);
  process.exit(1);
}

function check(label: string, condition: boolean): void {
  if (!condition) fail(`✗ ${label}`);
  console.log(`✓ ${label}`);
}

function reportErrors(label: string, result: TransformResult<unknown>): void {
  // Source data here is fixed and PII-free, so printing messages is safe.
  // Production adopters: `TransformError.message` can carry source values on the
  // Zod-validation path — see the README PII warning before copying this shape.
  if (result.errors.length > 0) {
    fail(
      `${label} produced errors: ${result.errors.map(e => `[${e.path ?? "?"}] ${e.message}`).join("; ")}`
    );
  }
}

function roundTrip(
  label: string,
  plugin: { schemas: { Opportunity: { toCommon?: unknown; fromCommon?: unknown } } },
  opts: { expectLegacyId: boolean }
): void {
  // The harness drives every plugin through one shape; the resolved common
  // output type read off a built plugin is fine here (this is test plumbing, not
  // the author-facing annotation the functions path above demonstrates).
  type HarnessCommon = z.output<typeof mappingsPlugin.schemas.Opportunity.commonSchema>;
  const opp = plugin.schemas.Opportunity;
  const toCommonFn = opp.toCommon as (s: GrantsGovSource) => TransformResult<HarnessCommon>;
  const fromCommonFn = opp.fromCommon as (c: HarnessCommon) => TransformResult<GrantsGovSource>;

  const common = toCommonFn(SOURCE_DATA);
  reportErrors(`${label} toCommon`, common);

  const back = fromCommonFn(common.result);
  reportErrors(`${label} fromCommon`, back);

  // The uuid round-trips verbatim in every path (unlike the parsed date fields).
  check(
    `${label}: opportunity_uuid round-trips`,
    back.result.data.opportunity_uuid === SOURCE_DATA.data.opportunity_uuid
  );
  // Three-state null survives both directions as a real null, not undefined.
  check(
    `${label}: source_url null ('doesn't apply') preserved`,
    back.result.data.source_url === null
  );
  // The legacy id only round-trips when a custom field carries it across.
  if (opts.expectLegacyId) {
    check(`${label}: legacy opportunity_id round-trips`, back.result.data.opportunity_id === 12345);
  }
}

function main(): void {
  console.log("=== Transforms via definePlugin ===\n");

  console.log("--- Path 1: declarative mappings ---");
  // `mappings` is kept on the entry for inspection (absent on the functions path).
  check("mappings inspectable", mappingsPlugin.schemas.Opportunity.mappings !== undefined);
  roundTrip("mappings", mappingsPlugin, { expectLegacyId: true });

  console.log("\n--- Path 2: hand-written functions ---");
  check(
    "mappings absent (functions path)",
    functionsPlugin.schemas.Opportunity.mappings === undefined
  );
  roundTrip("functions", functionsPlugin, { expectLegacyId: true });

  console.log("\n--- Path 3: no custom fields ---");
  check("customFields absent", noCustomFieldsPlugin.schemas.Opportunity.customFields === undefined);
  roundTrip("base", noCustomFieldsPlugin, { expectLegacyId: false });

  console.log("\n--- Path X: mappings XOR functions ---");
  demonstrateXorIsRejected();

  console.log("\n=== Example complete ===");
}

main();
